### ISSUE TITLE
Few remarks on the implement of `GM_xmlhttpRequest`

### ISSUE THREAD

**Issue by WhiteSevs** [JingMatrix/ChromeXt#104]
1. 希望不使用`fetch`来实现`GM_xmlhttpRequest`，因为有时候同源请求需要设置`headers`的`User-Agent`时，也是属于跨域了，使用`fetch`的话不会生效该`User-Agent`；
2. 提高兼容性，当`method`为`GET`时，如果`details`中存在`data`键，那么设置把`method`为`POST`，不然[https://github.com/JingMatrix/ChromeXt/blob/master/app/src/main/assets/GM.js](https://github.com/JingMatrix/ChromeXt/blob/master/app/src/main/assets/GM.js)的第`381`行会报错
```js
request = new Request(details.url, {
        cache: "force-cache",
        body: details.data, // 会报错
        ...details,
});
```

3. `GM_cookie`未实现，希望可以在脚本环境中删除该API，在·TamperMonkey·中的`GM_cookie`是`ojbect`类型；
4. 是否考虑在`iframe`内注入？`webview`可以`hook`接口`shouldInterceptRequest`，参数`WebResourceRequest `对它返回的html进行修改加入js；
5. `GM_xmlhttpRequest`似乎并没有对请求自动加入`Cookie`，比如我对一个api进行请求登录，返回的`headers`里有`Cookie`信息，后续的同域名请求并没有把`Cookie`加进去；

**Comment by JingMatrix** [comment]
Thanks for your interest in `ChromeXt` and valuable suggestions given here.

Your first two points are now implemented in  90ceb889e652503935c8706f69759eb3bb6a31fe. However, I shall keep using the `fetch` API since it can save the network-traffic when a UserScript is trying to get response data of current page resources. Moreover, it is faster than the Java native implement since there is a data conversion layer between Java and JavaScript.

For the third point, I shall add it to the dev plan and implement it soon.

For the fourth point, it is not an easy task for chromium based browsers, so it has lower priority in my plan. A possible implement requires [CDP](https://chromedevtools.github.io/devtools-protocol/). Everyone is welcome to contribute to this part. Please refer to the following codes if you are interested:
https://github.com/JingMatrix/ChromeXt/blob/90ceb889e652503935c8706f69759eb3bb6a31fe/app/src/main/java/org/matrix/chromext/devtools/WebSocketClient.kt#L70-L76

For the fifth point, I personally think that it is the responsibility of the UserScript to handle the relevant Cookie headers. This is simply because that my implement of `GM_xmlhttpRequest` is stateless: no data of previous responses are stored anywhere.

**Comment by WhiteSevs** [comment]
关于第一点，不光`User-Agent`，还有设置`Referer`、`Host`、`Origin`有时候也很重要的，不知道使用fetch能不能设置
我曾在代码中使用jQuery的`$.ajax`来代替过`GM_xmlhttpRequest`
```js
let headers_options_key = [
          "Accept-Charset",
          "Accept-Encoding",
          "Access-Control-Request-Headers",
          "Access-Control-Request-Method",
          "Connection",
          "Content-Length",
          "Cookie",
          "Cookie2",
          "Date",
          "DNT",
          "Expect",
          "Host",
          "Keep-Alive",
          "Origin",
          "Referer",
          "TE",
          "Trailer",
          "Transfer-Encoding",
          "Upgrade",
          "User-Agent",
          "Via",
        ]
```

**Comment by WhiteSevs** [comment]
> Thanks for your interest in `ChromeXt` and valuable suggestions given here.
> 
> Your first two points are now implemented in [90ceb88](https://github.com/JingMatrix/ChromeXt/commit/90ceb889e652503935c8706f69759eb3bb6a31fe). However, I shall keep using the `fetch` API since it can save the network-traffic when a UserScript is trying to get response data of current page resources. Moreover, it is faster than the Java native implement since there is a data conversion layer between Java and JavaScript.
> 
> For the third point, I shall add it to the dev plan and implement it soon.
> 
> For the fourth point, it is not an easy task for chromium based browsers, so it has lower priority in my plan. A possible implement requires [CDP](https://chromedevtools.github.io/devtools-protocol/). Everyone is welcome to contribute to this part. Please refer to the following codes if you are interested:
> 
> https://github.com/JingMatrix/ChromeXt/blob/90ceb889e652503935c8706f69759eb3bb6a31fe/app/src/main/java/org/matrix/chromext/devtools/WebSocketClient.kt#L70-L76
> 
> For the fifth point, I personally think that it is the responsibility of the UserScript to handle the relevant Cookie headers. This is simply because that my implement of `GM_xmlhttpRequest` is stateless: no data of previous responses are stored anywhere.

感谢修复，待会儿会去试一下最新版

**Comment by JingMatrix** [comment]
There is a [chroimum bug tracker](https://bugs.chromium.org/p/chromium/issues/detail?id=571722) for the `User-Agent` issue.
A priori, we are not sure which headers are not effective in the `fetch` API of chromium. Hence, I'd suggest that we change the condition for `fetch` API when new related bugs are reported. 
Currently, in your UserScript, are there other headers must be changed?

**Comment by WhiteSevs** [comment]
> There is a [chroimum bug tracker](https://bugs.chromium.org/p/chromium/issues/detail?id=571722) for the issue. A priori, we are not sure which headers are not effective in the API of chromium. Hence, I'd suggest that we change the condition for API when new related bugs are reported. Currently, in your UserScript, are there other headers must be changed?`User-Agent``fetch``fetch`

搜集了一下，我的脚本目前使用的`headers`有
```js
Accept
Authorization
Accept-Encoding
Accept-Language
Content-Type
Host
Origin
Pragma
Referer
x-csrf-token
X-Requested-With
```
另外关于这个有些建议[90ceb88](https://github.com/JingMatrix/ChromeXt/commit/90ceb889e652503935c8706f69759eb3bb6a31fe)`237`行，对`details.headers`中的`key`进行`小写/大写`转换判断，因为可能会有不规范写法，如`user-agent`或者`user-Agent`又或者`uSer-aGent`

**Comment by JingMatrix** [comment]
Now `ChromeXt` will take all [forbidden headers listed on MDN](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name) into consideration.
Moreover, you can specify the `foreceCORS` option, see b7b023e9331564ba1165e409e8e36a3f833ca973 .

No worry about the upper or lower spelling cases of the header keys, the `Header` API of JavaScript can take care of them automatically.

**Comment by WhiteSevs** [comment]
刚刚发现一个新bug，当用户的`details`期望为`json`时，此刻的返回的`response.responseText`内其实是`html`，这时候会解析失败
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/2ef0ec21-3c2f-4dbd-88cd-363daf504fd5)

**Comment by JingMatrix** [comment]
Could you give an exmaplar URL for the above bug? Maybe you can missed a `content-type` header.

**Comment by WhiteSevs** [comment]
[https://up.woozooo.com/mlogin.php](https://up.woozooo.com/mlogin.php)
用于`蓝奏云`网盘登录
👇是`details`结构
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/830a9e01-06f5-423a-956f-5fa6da858eb0)

**Comment by JingMatrix** [comment]
I cannot reproduce it using `cURL`.
Even without any specific headers, the response is still a JSON string:
```sh
curl -v 'https://up.woozooo.com/mlogin.php' --data-raw 'task=3&uid=jingmatrix%40gmail.com&pwd=test&setSessionId=&setSig=&setScene=&setToken=&formhash=ab2489d6'
```
The Java implement of `GM_xmlhttpRequest` should be the same as `cURL`.
Could you also show the response data? Maybe you can expand the promise, and find the response data.

**Comment by WhiteSevs** [comment]
我使用TamperMonkey的请求复现了一下，事实上是`reponseText`不是`json`数据，它的`response`直接为`undefined`
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/900ab898-0f59-4c19-8549-09ce519c8320)

**Comment by JingMatrix** [comment]
The `finalUrl` indicates that your request didn't follow the redirect correctly. You may retry with the URL `https://up.woozooo.com/mlogin.php`.

If `TamperMonkey` cannot response `JSON` data, then it seems not to be a bug of script manager.

**Comment by WhiteSevs** [comment]
该请求为302重定向跳转到了`https://up.woozooo.com/myfile.php`，导致了`response.responseText`内容应该是`json`变成了`html`，
·TamperMonkey·对`response.reponse`做的处理是不进行`JSON.parse`，`ChromeXt`也需要在上面进行兼容性处理，不然`ChromeXt`的`data.response = JSON.parse(data.responseText);`就会执行失败导致无法调用脚本的`onloadCallBack`

> 刚刚发现一个新bug，当用户的期望为时，此刻的返回的内其实是，这时候会解析失败`details.responseType``json``response.responseText``html` ![image](https://user-images.githubusercontent.com/50544447/268224065-2ef0ec21-3c2f-4dbd-88cd-363daf504fd5.png)

**Comment by JingMatrix** [comment]
It is wired about the redirection, because `cURL` tells me that `https://up.woozooo.com/myfile.php` redirects to `./mlogin.php`, but you claimed the reverse.

I think when you specify `responseType`, you are expecting a `JSON` reponse. And if such response is not available, an error should be thrown.
Hence, I will reject the Promise (after catching it in JSON.parse) with unparsed response. Do you think it is more reasonable?

**Comment by WhiteSevs** [comment]
很合理，但是如果resolve的话兼容性会更好一些

> It is wired about the redirection, because `cURL` tells me that `https://up.woozooo.com/myfile.php` redirects to `./mlogin.php`, but you claimed the reverse.
> 
> I think when you specify `responseType`, you are expecting a `JSON` reponse. And if such response is not available, an error should be thrown. Hence, I will reject the Promise (after catching it in JSON.parse) with unparsed response. Do you think it is more reasonable?

**Comment by JingMatrix** [comment]
In the commit aba8e9dd1280722b71335f71475f5885e1b856c6, I throw the parsing error out. Please tell me if this solution is acceptable for you. I think that unless you are using the async version `GM.xmlHttpRequest`, your code won't stop when there is a promise error.

Also, I implemented the basic APIs of `GM_cookie` in 567dcddb81046d1bcec564aa6f65a6b7ac3a13ed, but I am not very familar with the usage case of it. Could you please give some usage scenarios? So that I am more aware of what kind of functionalities should be included.

**Comment by JingMatrix** [comment]
For your fifth point, I found a way to implement it in `Java`. It was my ignorance to not implement it.

**Comment by WhiteSevs** [comment]
> In the commit [aba8e9d](https://github.com/JingMatrix/ChromeXt/commit/aba8e9dd1280722b71335f71475f5885e1b856c6), I throw the parsing error out. Please tell me if this solution is acceptable for you. I think that unless you are using the async version `GM.xmlHttpRequest`, your code won't stop when there is a promise error.
> 
> Also, I implemented the basic APIs of `GM_cookie` in [567dcdd](https://github.com/JingMatrix/ChromeXt/commit/567dcddb81046d1bcec564aa6f65a6b7ac3a13ed), but I am not very familar with the usage case of it. Could you please give some usage scenarios? So that I am more aware of what kind of functionalities should be included.

1. 试了一下相较之前的并没太大区别，虽然在`ChromeXt`内抛出了`JSON.parse`错误，但是脚本的`onload`并未触发，我删除了自己脚本的`responseType: "json"`，自己在`onload`返回内对`response.responseText`进行了处理。
2. 下面是一些我使用的`GM_cookie`的例子
```js
// 用来获取用户是否登录的Cookie，该Cookie是HttpOnly，document.cookie无法获取到
function getCookie(cookieName) {
        return new Promise((resolve) => {
          GM_cookie.list({ name: cookieName }, function (cookies, error) {
            if (error) {
              resolve(null);
            } else {
              if (cookies.length == 0) {
                resolve(null);
              } else {
                resolve(cookies[0].value);
              }
            }
          });
        });
      }

await getCookie("userLogin")
```

**Comment by WhiteSevs** [comment]
我又找到一个关于`response.responseText`编码问题
```js
GM_xmlhttpRequest({
    url:"https://tieba.baidu.com/f/search/res?isnew=1&kw=%C4%E6%CB%AE%BA%AE%CA%D6%D3%CE&qw=%CE%E8%D1%F4%B3%C7&un=&rn=10&pn=0&sd=&ed=&sm=1",
    method:"get",
    headers: {
    Referer: "https://tieba.baidu.com/f?ie=utf-8&kw=%E9%80%86%E6%B0%B4%E5%AF%92%E6%89%8B%E6%B8%B8",
    Host: "tieba.baidu.com",
    Accept:
      "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
  },
  responseType: "html",
    onload:(resp)=>{console.log(resp)}
})
```
## 在`TamperMonkey`中正常解码👇
![{11014FD9-1D3E-4726-98AB-2FCAC15DD368}](https://github.com/JingMatrix/ChromeXt/assets/50544447/46eef79d-faaf-4e60-b82a-586bae2f76ca)

## 在`ChromeXt`中乱码👇
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/9378e957-a685-4033-82ae-fe859f1290d8)

猜测是`content-type: text/html; charset=GBK`的缘故？

**Comment by JingMatrix** [comment]
Thanks for the feedbacks!
1. Now parsing error won't block `onload`.
2. To get `httpOnly` cookie, I should use [CDP](https://chromedevtools.github.io/devtools-protocol/1-3/Network/#type-Cookie). This can be done later.
3. You are right, I was assuming `UTF-8` encoding. Should be fixed soon.

**Comment by WhiteSevs** [comment]
> Thanks for the feedbacks!
> 
> 1. Now parsing error won't block `onload`.
> 2. To get `httpOnly` cookie, I should use [CDP](https://chromedevtools.github.io/devtools-protocol/1-3/Network/#type-Cookie). This can be done later.
> 3. You are right, I was assuming `UTF-8` encoding. Should be fixed soon.

关于第2点，我在`TamperMonkey`中测试了一下，当我在`headers`里设置了`{cookie:"xxxxxxx"}`，它会自动将属于该`domain`的为`HttpOnly`的Cookie自动添加到`Cookie`前面
例如，当前`domain`下的`cookie`有

|key | value |  domain| path | expires | ... | HttpOnly| 
| --- | --- | --- |--- |--- |--- |--- |
|github_test| 1|github.com | / | ....| | |
|github_test2| 2|github.com | / | ....| | √|

```js
// 我设置的cookie

headers: {
  "user-agent":".....",
   cookie: "github_test=1;",
}
```
发出去的Cookie会变成
`Cookie: github_test2=2;github_test=1`
也就是，Cookie前面的是当前的，后面的属于用户设置的。
当我尝试通过抓包通过`lspatch`打包的`ChromeXt`时，发现请求头存在多个`Cookie键`，且我设置的`Cookie`是小写的`cookie`

![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/56aa2779-7669-4d68-ac61-8c321ef0b96b)

**Comment by JingMatrix** [comment]
`httpOnly` cookie is not supported yet, but will be done soon. Once it is implemented, `httpOnly` cookie headers will be appended.
As for the given screenshot, is it a `CORS` request? 
Could you please describe what should be the expected behavior?
If I am correct, in the API, `cookie` should be set as a property of `details` instead of `headers`.

**Comment by WhiteSevs** [comment]
> `httpOnly` cookie is not supported yet, but will be done soon. Once it is implemented, `httpOnly` cookie headers will be appended. As for the given screenshot, is it a `CORS` request? Could you please describe what should be the expected behavior? If I am correct, in the API, `cookie` should be set as a property of `details` instead of `headers`.

对，是跨域请求，修改了`user-agent`，该请求的作用是发送一个签到请求，需要`Cookie`验证当前身份是否和url中的`formhash`匹配，`cookie`确实应该是放在`details`中而非`headers`内

**Comment by WhiteSevs** [comment]
> Now `httpOnly` support is implemented. Please test it and share your feedbacks! 😃

我尝试了一下最新的[https://github.com/JingMatrix/ChromeXt/actions/runs/6228644262](https://github.com/JingMatrix/ChromeXt/actions/runs/6228644262)，似乎有严重的问题，`GM_xmlhttpRequest`的请求不触发任何回调
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/88043447-5b6a-4074-a587-924144d69e7d)

**Comment by JingMatrix** [comment]
Cannot reproduce your issue, in my devices,
```javascript
GM_xmlhttpRequest({
    url: "https://bbs.binmt.cc/k_misign-sign.html?operation=qiandao&format=button&formhash=TESTHASHVALUE&inajax=1&ajaxtarget=midaben_sign",
    onload: (r)=>console.log(r.response),
    onerror: (r)=>console.log(r),
    timeout: 5000,
    forceCORS: true
})
```
returns a response with `您当前的访问请求当中含有非法字符，已经被系统拒绝`.

**Comment by WhiteSevs** [comment]
> Cannot reproduce your issue, in my devices,
> 
> ```js
> GM_xmlhttpRequest({
>     url: "https://bbs.binmt.cc/k_misign-sign.html?operation=qiandao&format=button&formhash=TESTHASHVALUE&inajax=1&ajaxtarget=midaben_sign",
>     onload: (r)=>console.log(r.response),
>     onerror: (r)=>console.log(r),
>     timeout: 5000,
>     forceCORS: true
> })
> ```
> 
> returns a response with `您当前的访问请求当中含有非法字符，已经被系统拒绝`.


使用这个仍是未触发

https://github.com/JingMatrix/ChromeXt/assets/50544447/52bf210e-d024-484d-abd2-1fc351dc25db

**Comment by JingMatrix** [comment]
I still cannot reproduce it with `Via`. Maybe you need to set breakpoints to see why this part of code
https://github.com/JingMatrix/ChromeXt/blob/523eed33da5112577700f5e5cabaa991790c4605/app/src/main/assets/GM.js#L493-L496
is not triggered.

The `onload` function should be trigger by this call:
https://github.com/JingMatrix/ChromeXt/blob/523eed33da5112577700f5e5cabaa991790c4605/app/src/main/assets/GM.js#L793

**Comment by WhiteSevs** [comment]
我找到bug了，并不是这儿，而是`await GM_cookie.list()`，它并没有返回，所以一直处于等待中。
https://github.com/JingMatrix/ChromeXt/blob/523eed33da5112577700f5e5cabaa991790c4605/app/src/main/assets/GM.js#L625

跟进
https://github.com/JingMatrix/ChromeXt/blob/523eed33da5112577700f5e5cabaa991790c4605/app/src/main/assets/GM.js#L145-L147

似乎监听未触发
https://github.com/JingMatrix/ChromeXt/blob/523eed33da5112577700f5e5cabaa991790c4605/app/src/main/assets/GM.js#L120-L121

https://github.com/JingMatrix/ChromeXt/assets/50544447/9a37cbb8-9733-483b-88d8-c448c559f468

**Comment by JingMatrix** [comment]
Thanks for reporting this bug, it is fixed now. This is a bug about thread safe usage of WebView class, see details in 666253121bde33669cf9679ac66c1c427dba3d98.

**Comment by WhiteSevs** [comment]
> Thanks for reporting this bug, it is fixed now. This is a bug about thread safe usage of WebView class, see details in [6662531](https://github.com/JingMatrix/ChromeXt/commit/666253121bde33669cf9679ac66c1c427dba3d98).

刚刚尝试了最新版，似乎未修复该问题

**Comment by JingMatrix** [comment]
> 刚刚尝试了最新版，似乎未修复该问题

That seems impossible, on my device, `GM_bridge.GM_cookie.list().then(a => console.log(a))` returns correctly the cookies of `Via`. Maybe, your `Via` is not fully restarted?

**Comment by WhiteSevs** [comment]
是的，现在可以工作了，应该是我没重启`via`的问题，在测试签到请求时，对比发现，Cookie的添加方式似乎有问题。
存在两个Cookie在请求头中，且，第一个`Cookie`中的`cQWy_2132_saltkey=Nlp...`，这个是当前真正的，第二个`Cookie`中的`cQWy_2132_saltkey=toW...`是错误的，猜测是之前的？
![{CF1DCD82-7645-408f-9106-49272FDBE7B2}](https://github.com/JingMatrix/ChromeXt/assets/50544447/b6c3edd4-a4e1-446a-aa85-defef189b3a0)

且发出去的`Cookie`还有缺失，使用`GM_cookie.list`和`DevTools`中进行比对是完全一致的
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/7c735866-23cb-4c7b-b98a-cf6b505835df)

**Comment by JingMatrix** [comment]
Yes, you are right. The problem in `Java` part can be tricky.
By the way, which part of cookies is misssing from the first `Cookie` header?
`GM_cookie.export(location.origin)` returns the adds `HttpOnly` cookie.

**Comment by WhiteSevs** [comment]
第一个Cookie中缺失所有非`HttpOnly`的`Cookie`，所以服务器无法验证当前请求的身份。
`GM_cookie.export(location.origin)`存在问题，返回空值
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/5299e84e-b28a-467c-be6d-393b8f48e4fe)

**Comment by JingMatrix** [comment]
I suppose that we should only add `httpOnly` cookies since they are the ones not accessible by `document.cookie`, isn't that true? Does `TamperMonkey` include all of them?

**Comment by WhiteSevs** [comment]
`TamperMonkey `会处理所有的`Cookie`，包括`httpOnly`和`document.cookie`的，因此使用者大部分情况下不太关注`Cookie`的使用问题

**Comment by JingMatrix** [comment]
For WebView browsers, the cookie management should be Okay now.

However, for Chromium based browsers, the UserScript should take care of `Set-Cookie` headers if the request has the same domain of current page.

**Comment by JingMatrix** [comment]
There was a mistake in the previous commit. Now the cookie problem should be solved gracefully.

**Comment by WhiteSevs** [comment]
我试了一下，签到功能正常了，还有其它问题，比如，当前我在`https://bbs.binmt.cc`使用API向`https://www.z4a.net/`图床发送了登录请求，登录成功，但是`ChromeXt`并未存储相关`Cookie`？后面的上传图片就验证`Cookie`失败了，然后打开新标签也进`https://www.z4a.net/`图床进行登录账号，再返回`https://bbs.binmt.cc`内上传图片，这时候验证`Cookie`成功了，也就是说，通过`GM_xmlhttpRequest`无法进行这种`跨域名登录`的操作。

在`TamperMonkey`的流程为👇
`当前在域名bbs,binmt.cc下` => `跨域登录www.z4a.net` => `自动保存登录www.z4a.net的Cookie` => `上传图片` => `获取图片的结果`

**Comment by WhiteSevs** [comment]
可以使用这个脚本[https://greasyfork.org/zh-CN/scripts/475722-greasyfork%E4%BC%98%E5%8C%96](https://greasyfork.org/zh-CN/scripts/475722-greasyfork%E4%BC%98%E5%8C%96)，用于在`greasyfork.org`上的，自动登录账号，`录入账号`功能在菜单中

**Comment by JingMatrix** [comment]
The function you asked is the same as [tracking](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies#tracking_protection).
It might be that the (WebView) browser has it own setting against various tracking. Changing this setting might incure blames from users, so I prefer not to change it.
Hence, I suggest you to store the cookie in your own scripts only and send them by setting the `cookie` parameter.
`ChromeXt` provides two helper funtions for you to parse headers and export cookies:
https://github.com/JingMatrix/ChromeXt/blob/9e6454f5f1d157cd760ab01bc6f19a4aba29bb65/app/src/main/assets/GM.js#L124-L132
and
https://github.com/JingMatrix/ChromeXt/blob/9e6454f5f1d157cd760ab01bc6f19a4aba29bb65/app/src/main/assets/GM.js#L734-L737

**Comment by WhiteSevs** [comment]
> The function you asked is the same as [tracking](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies#tracking_protection). It might be that the (WebView) browser has it [own setting](https://developer.android.com/reference/android/webkit/CookieManager#setAcceptThirdPartyCookies(android.webkit.WebView,%20boolean)) against various tracking. Changing this setting might incure blames from users, so I prefer not to change it. Hence, I suggest you to store the cookie in your own scripts only and send them by setting the `cookie` parameter. `ChromeXt` provides two helper funtions for you to parse headers and export cookies:
> 
> https://github.com/JingMatrix/ChromeXt/blob/9e6454f5f1d157cd760ab01bc6f19a4aba29bb65/app/src/main/assets/GM.js#L124-L132
> 
> 
> and
> https://github.com/JingMatrix/ChromeXt/blob/7633fe3baa1fcfb6287e002e2c7e6bfb3dd21032/app/src/main/assets/GM.js#L734-L737

这样的话似乎可以`GM_cookie.list`然后`GM_cookie.export`来获取`Cookie`放到`details`里，但是不能获取其它域名的`Cookie`

**Comment by WhiteSevs** [comment]
我在图床登录，然后使用`GM_cookie.export`导出时发现，它的有个`cookie`是`.z4a.net`，多了一个`.`，`GM_cookie.export`后就缺失了这个值
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/d5880823-63c5-4ea6-b9bd-e6af7fa793f0)

**Comment by JingMatrix** [comment]
To obtain cookies from other domain,  you need to parse the response headers of an xhr request in your `onload` function. See for example:
https://github.com/JingMatrix/ChromeXt/blob/41d61a84467735aa651c54d8a21043a5317446c8/app/src/main/assets/GM.js#L810-L820
, where you call the static method `ResponseSink.parseCookie`.
The above example stores cookies in _current domain_ for _Chromium_ based browsers.
You can then export it using `GM_info.export(url, cookies)`.

I searched some information about the leading `.` of domain and realized that it should be included.Now it is fixed 5by a git force push°.

**Comment by JingMatrix** [comment]
I am working on issue #115 to make `GM_xhr` stateful. Your goal should be reached when the implement is done. It should behave as `TamperMonkey`.

**Comment by JingMatrix** [comment]
The re-implement of `GM_cookie` is done.
Now, it should work in the same way as `TamperMonkey`.

**Comment by JingMatrix** [comment]
@WhiteSevs Did you succeed to run your script with the latest commits of `ChromeXt`? I am planning to release a new version of ChromeXt this weekend. If you find no errors, please drop a message here. Also, this issue will be closed after that.

**Comment by WhiteSevs** [comment]
> @WhiteSevs Did you succeed to run your script with the latest commits of `ChromeXt`? I am planning to release a new version of ChromeXt this weekend. If you find no errors, please drop a message here. Also, this issue will be closed after that.

ok，现在基本上除了那些需要跨域登录的操作都没问题了

**Comment by JingMatrix** [comment]
CORS login should be working now. Did you still fail to do that?

**Comment by WhiteSevs** [comment]
> CORS login should be working now. Did you still fail to do that?

Yes，抓包看`登录`请求成功后`301重定向`，该重定向请求并没有携带`Cookie`

**Comment by JingMatrix** [comment]
If you receive the `301` redirection, the cookie won't be sent according to [this explanation](https://stackoverflow.com/a/73599289/7870953). You should change your request URL to the redirected one and try again.

**Comment by WhiteSevs** [comment]
> If you receive the `301` redirection, the cookie won't be sent according to [this explanation](https://stackoverflow.com/a/73599289/7870953). You should change your request URL to the redirected one and try again.

设置`redirect: "manual"`可以正常，但是在`TamperMonkey`中却是不正常的。
在上传图片时出现错误👇
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/1fdb8008-8510-42e1-81aa-7c355ad2e9c9)

**Comment by JingMatrix** [comment]
Currently in `ChromeXt`, `redirect: "manual"` is the same as `redirect: "follow"`, there should be no difference.
Could you give the details of you URL request? I may try it on my side.
`500 error` means the error is not in the client side.

**Comment by WhiteSevs** [comment]
```js
// token可以在www.z4a.net页面HTML中搜索到 => PF.obj.config.auth_token
GM_xmlhttpRequest({
    url:"https://www.z4a.net/login",
    method:"POST",
    redirect: "manual",
    data: `login-subject=${user}&password=${pwd}&auth_token=${xxxxx}`,
    headers: {
              "User-Agent": GM_bridge.Utils.getRandomPCUA(),
            },
    onload:(r)=>{console.log(r)},
    onerror:(r)=>{console.log(r)}
})

```

**Comment by JingMatrix** [comment]
Okay, I see. That is because I didn't process `FormData` in `GM_xmlhttpRequest` yet. This will be done soon. It seems not to be a problem of `cookie`.

**Comment by WhiteSevs** [comment]
> ```js
> redirect: "manual",
> ```

对，但是上传的前一个步骤是先登录，那个步骤有重定向Cookie问题

**Comment by JingMatrix** [comment]
> 对，但是上传的前一个步骤是先登录，那个步骤有重定向Cookie问题

Don't worry, if you login successfully, `ChromeXt` will take good care of the cookies.

**Comment by JingMatrix** [comment]
Currently, there is no way to stop `Java` from redirection if `HttpURLConnection` is used. This is a Android bug. I may need to re-implement Http Connections.

**Comment by JingMatrix** [comment]
@WhiteSevs  I have tested another website `https://imgse.com/`, the uploding is working well with the latest commit of ChromeXt.

**Comment by JingMatrix** [comment]
@WhiteSevs Did you try the cookie login again?
I am about to publish a new release of `ChromeXt` today. Currently, at least on my devices, it is working.

**Comment by WhiteSevs** [comment]
> @WhiteSevs Did you try the cookie login again? I am about to publish a new release of `ChromeXt` today. Currently, at least on my devices, it is working.

貌似有点问题，在登录时返回的`responseText`内没有内容
![image](https://github.com/JingMatrix/ChromeXt/assets/50544447/d465aa04-fd45-4c9f-8e50-0a2ae39cf67c)

**Comment by JingMatrix** [comment]
That is because the default value for `redirect` is `manual`, and the `status` is `301`.
You will see some response if you set it to `follow`.
This redirection is not important regarding to login.

**Comment by JingMatrix** [comment]
Do you need the response for your script? If not, you can safely ignore it for a while.
I will change the code for redirection later.
How about the upload? Did it work as expected?

**Comment by WhiteSevs** [comment]
是的，我脚本中有段是根据这个responseText内容来判断是否已成功登录，这个报错很奇怪，我是用的登录，并没有使用上传

**Comment by JingMatrix** [comment]
The redirect was due to the [unclear specification of 301](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections#permanent_redirections), i.e., whether to change the method from `POST` to `GET`.
I will follow the design of chromium to do the modification. This will be fixed soon.

To decide whether the login is succeeded, you can simply test if the `status` is `200 or 301`. There is no need to consider the response.

**Comment by JingMatrix** [comment]
Now I have overwritten `POST` to `GET` for `301` redirections.
Hopefully, nobody in the future will need to change this behavior.

**Comment by JingMatrix** [comment]
I have change the default value of `redirect` to be `follow`.
Note that in `ChromeXt`, you can use the `onredirect` parameter to listen for the `redirect` event.

**Comment by WhiteSevs** [comment]
奇怪，最新的版本无法进行图床登录，上个版本我测试了下还可以进行登录且上传图片

**Comment by JingMatrix** [comment]
Cannot reproduce on my phone. I tested with the following code, and it worked as expected.
```javascript
let url = "https://imgse.com/";

async function get_token() {
  const html = await GM_bridge.GM_xmlhttpRequest({ url, forceCORS: true });
  return html.match(/PF.obj.config.auth_token = "(\w+)"/)[1];
}

function login(
  auth_token = PF.obj.config.auth_token,
  redirect = "follow",
  nocache = true
) {
  const form = new FormData();
  form.append("login-subject", user);
  form.append("password", pwd);
  form.append("auth_token", auth_token);
  return GM_bridge.GM_xmlhttpRequest({
    url: url + "login",
    method: "POST",
    redirect,
    nocache,
    onredirect: (xhr) => console.log(xhr),
    data: form,
    onload: (xhr) => console.log(xhr.status),
  });
}

function createForm() {
  if (document.querySelector("#chromext")) return;
  const input = document.createElement("input");
  input.setAttribute("type", "file");
  input.setAttribute("accept", "image/*");
  input.id = "chromext";
  document.body.prepend(input);
  return input;
}

function upload(auth_token = PF.obj.config.auth_token, putFile = true) {
  input = document.querySelector("#chromext");
  if (putFile && typeof input.files[0] == "undefined") {
    console.warn("No file chosen yet");
  }
  const form = new FormData();
  form.append("type", "file");
  form.append("action", "upload");
  form.append("timestamp", new Date().getTime());
  if (putFile) form.append("source", input.files[0]);
  form.append("auth_token", auth_token);
  form.append("nsfw", 0);
  GM_bridge.GM_xmlhttpRequest({
    responseType: "json",
    url: url + "json",
    method: "POST",
    data: form,
    timeout: 30000,
    onload: (xhr) => console.log(xhr.response),
  });
}

function test() {
  const promise = get_token().then((token) => {
    console.log("Auth_token:", token);
    login(token);
  });
  createForm();
  return promise;
}
```
You run it with `test()` and then `upload(token)` with the token given in the console.

**Comment by WhiteSevs** [comment]
`imgse.com`倒是可以，在测试`z4a.net`登录时，发现没加`redirect: "follow"`登录会失败，然后加上后，就登录成功了，再刷新DevTools，上传功能就正常了

**Comment by JingMatrix** [comment]
Even with `z4a.net`, it is not reproducible on my device. 
It is most likely to be a caching issue, or your traffic is somehow blocked by `z4a.net`.
To avoid such mistakes, please ensure to restart your browser each time when you test it.

This will not be considered as a bug. If you are Okay with that, please close the issue.

**Comment by WhiteSevs** [comment]
OK，删除脚本后重新安装了下，第一次登录失败，第二次登录成功且上传图片完毕，应该是这个图床网站的问题
